makefiles/docker.inc.mk

```diff
@@ -19,6 +32,25 @@ else
   export INSIDE_DOCKER := 0
 endif
 
+ifeq (0,$(INSIDE_DOCKER))
+  ifeq ($(DOCKER_IMAGE),$(DOCKER_IMAGE_DEFAULT))
+    IMAGE_PRESENT:=$(shell $(DOCKER) image inspect $(DOCKER_IMAGE) 2>/dev/null >/dev/null && echo 1 || echo 0)
+    ifeq (0,$(IMAGE_PRESENT))
+      $(warning Required docker image $(DOCKER_IMAGE) not installed)
+      ifeq (1,$(DOCKER_AUTO_PULL))
+        $(info Pulling required image now automatically. You can disable this with DOCKER_AUTO_PULL=0)
+        $(shell $(DOCKER) pull $(DOCKER_REPO)@sha256:$(DOCKER_TESTED_IMAGE_REPO_DIGEST))
```

Comment:
For some reason this bails out on my system with

    /home/mikolai/TUD/Code/RIOT/makefiles/docker.inc.mk:42: *** target pattern contains no '%'.  Stop.

Seems to be connected to the usage of `@`.

Reply:
That error message is pretty crazy, as if make was parsing the arguments to `$(shell ...)` as a make target.

Did quoting it help?

Reply:
Which version of make are you using btw.? The one from latest Ubuntu LTS?

Reply:
Humm I tried

```
$(shell $(DOCKER) pull "$(DOCKER_REPO)@sha256:$(DOCKER_TESTED_IMAGE_REPO_DIGEST)")
```

still the same :/

This is `GNU Make 4.3`.

Reply:
And now?

Reply:
In a debian podman container I ran:

```
$ DOCKER=podman make BUILD_IN_DOCKER=1
/home/maribu/Repos/software/RIOT/master/makefiles/docker.inc.mk:39: Required docker image sha256:f5951bc41dfface6cac869181d703e62cbdd3b7976b0946130a38f2e658000b3 not installed
Pulling required image now automatically. You can disable this with DOCKER_AUTO_PULL=0
Trying to pull docker.io/riot/riotbuild@sha256:75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1...
Getting image source signatures
Copying blob 641b60a8ef3b done  
```

In that container, I also have

```
$ make --version
GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

Reply:
Fun, works for me too with `DOCKER=podman`, but not with `DOCKER=docker` (or no env variable). :confused:

Reply:
I'm really out of ideas on how to debug this... :/

Reply:
I sadly cannot install docker in a container. I did try it again with a debian stable and podman-docker (that is, podman provides the docker command). That did just work. (And according to the error message the error should occur prior to podman or docker being invoked.)

Could you check if `env -i PATH="$PATH" make BUILD_IN_DOCKER=1` works for you? Maybe there is some environment variable that somehow affects the process that I do not have in my freshly set up debian container?

Reply:
> Could you check if `env -i PATH="$PATH" make BUILD_IN_DOCKER=1` works for you? Maybe there is some environment variable that somehow affects the process that I do not have in my freshly set up debian container?

Sadly still the same problem.

Reply:
FWIW, I can reproduce with a minimal example: `make all` on this Makefile:

```
$(shell docker pull docker.io/riot/riotbuild@sha256:75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1)

all:
```

Reply:
Does it work if you do

```Makefile
IGNORED := $(shell ...)
```

instead of

```Makefile
$(shell ...)
```

?

Reply:
This however works:

```
TEST := $(shell docker pull docker.io/riot/riotbuild@sha256:75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1)

all:
```

Reply:
I just added the `docker-pull` make goal and add this as dependency to the command running docker if and only if the image is locally not available and automatic pulling is enabled.

Reply:
That works for me! :tada: